# Load data

In [190]:
import pandas as pd

In [191]:
df=pd.read_csv("../data/train.csv")

In [192]:
df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [193]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [194]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB



###Insight
1. Age has 714 non-null. 177 values are missing. we need to handle them 
2. sex is object. we need to convert categorical into numbers.

In [195]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [196]:
df['Survived'].value_counts()

Survived
0    549
1    342
Name: count, dtype: int64

###Insights
1. Age has missing values
2. Fare has outliers
3. Pclass and Age are important
4. most passengers were young adults, travled alone, were in lower class

In [197]:
df.groupby('Sex')['Survived'].mean()

Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64

Female passengers had a survival rate nearly 4x higher than males, indicating strong prioritization during evacuation

In [198]:
df.groupby('Pclass')['Survived'].mean()

Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64

Passengers in 1st class had a survival rate of 63%, compared to only 24% in 3rd class, highlighting the impact of socioeconomic status on survival

In [199]:
df.groupby(['Sex','Pclass'])['Survived'].mean()

Sex     Pclass
female  1         0.968085
        2         0.921053
        3         0.500000
male    1         0.368852
        2         0.157407
        3         0.135447
Name: Survived, dtype: float64

Survival depended on both gender and class, but gender had the strongest effect and Class amplified or reduced that effect

# Data Cleaning 

Check Missing Values

In [200]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Cabin has most number of missing values
Age also have missing values
embarked has few msising values

##Handle Missing values

In [201]:
df['Age']=df['Age'].fillna(df['Age'].median())

In [202]:
df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,28.0,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [203]:
df['Age']

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888    28.0
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

In [204]:
df['Embarked']=df['Embarked'].fillna(df['Embarked'].mode()[0])

### Droping Cabin column

- 77% data is missing (687/891)
- Too much to recover meaningfully
- Will add noise, not value

In [205]:
df.drop('Cabin',axis=1,inplace=True)

##Re-Checking missing values

In [206]:
df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

# Feature engineering

In [207]:
df['FamilySize']=df['SibSp']+df['Parch']+1

In [208]:
df['IsAlone']=(df['FamilySize']==1).astype(int)

In [209]:
df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,FamilySize,IsAlone
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S,2,0
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,2,0
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,1,1
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,2,0
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S,1,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,S,1,1
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,S,1,1
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,28.0,1,2,W./C. 6607,23.4500,S,4,0
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C,1,1


Title

In [210]:
# df['Title']=df['Name'].str.extract(' ([A-Za-z]+)\.', expand=False)

In [211]:
# df['Title'] = df['Title'].replace(['Mlle','Ms'], 'Miss')
# df['Title'] = df['Title'].replace(['Mme'], 'Mrs')

In [212]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,FamilySize,IsAlone
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S,2,0
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,2,0
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,1,1
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,2,0
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S,1,1


Removing Name, Ticket, PassengerId as they are not useful

In [213]:
df.drop(['Name','Ticket','PassengerId'],axis=1,inplace=True)

In [214]:
# df['AgeGroups']=pd.cut(df['Age'], bins=5)

In [215]:
df.head()

,Survived,Pclass,Sex,Age,SibSp,Parch,Fare,Embarked,FamilySize,IsAlone
0,0,3,male,22.0,1,0,7.2500,S,2,0
1,1,1,female,38.0,1,0,71.2833,C,2,0
2,1,3,female,26.0,0,0,7.9250,S,1,1
3,1,1,female,35.0,1,0,53.1000,S,2,0
4,0,3,male,35.0,0,0,8.0500,S,1,1


## Encode categorical variables

In [216]:
df.head()

,Survived,Pclass,Sex,Age,SibSp,Parch,Fare,Embarked,FamilySize,IsAlone
0,0,3,male,22.0,1,0,7.2500,S,2,0
1,1,1,female,38.0,1,0,71.2833,C,2,0
2,1,3,female,26.0,0,0,7.9250,S,1,1
3,1,1,female,35.0,1,0,53.1000,S,2,0
4,0,3,male,35.0,0,0,8.0500,S,1,1


In [217]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Survived    891 non-null    int64  
 1   Pclass      891 non-null    int64  
 2   Sex         891 non-null    object 
 3   Age         891 non-null    float64
 4   SibSp       891 non-null    int64  
 5   Parch       891 non-null    int64  
 6   Fare        891 non-null    float64
 7   Embarked    891 non-null    object 
 8   FamilySize  891 non-null    int64  
 9   IsAlone     891 non-null    int32  
dtypes: float64(2), int32(1), int64(5), object(2)
memory usage: 66.3+ KB


1. Converting Categorical columns to numeric

In [218]:
df['Sex']

0        male
1      female
2      female
3      female
4        male
        ...  
886      male
887    female
888    female
889      male
890      male
Name: Sex, Length: 891, dtype: object

In [219]:
df['Sex']=df['Sex'].map({'male':0,'female':1})

In [220]:
df['Sex']

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    1
889    0
890    0
Name: Sex, Length: 891, dtype: int64

In [221]:
df['Embarked'].value_counts()

Embarked
S    646
C    168
Q     77
Name: count, dtype: int64

In [222]:
df=pd.get_dummies(df, columns=['Embarked'],drop_first=True)

In [223]:
# df['Title'].value_counts()

In [224]:
#df = pd.get_dummies(df, columns=['Title'], drop_first=True)

In [225]:
df

,Survived,Pclass,Sex,Age,SibSp,Parch,Fare,FamilySize,IsAlone,Embarked_Q,Embarked_S
0,0,3,0,22.0,1,0,7.2500,2,0,False,True
1,1,1,1,38.0,1,0,71.2833,2,0,False,False
2,1,3,1,26.0,0,0,7.9250,1,1,False,True
3,1,1,1,35.0,1,0,53.1000,2,0,False,True
4,0,3,0,35.0,0,0,8.0500,1,1,False,True
...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,0,27.0,0,0,13.0000,1,1,False,True
887,1,1,1,19.0,0,0,30.0000,1,1,False,True
888,0,3,1,28.0,1,2,23.4500,4,0,False,True
889,1,1,0,26.0,0,0,30.0000,1,1,False,False


# Train/test split

In [226]:
X=df.drop('Survived',axis=1)


In [227]:
y=df['Survived']

In [228]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train model

In [229]:
from sklearn.ensemble import RandomForestClassifier
model=RandomForestClassifier(random_state=42)
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

Make predictions

In [230]:
y_pred=model.predict(X_test)

In [231]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

print('Accuracy',accuracy_score(y_test, y_pred))


Accuracy 0.8268156424581006


In [232]:
print('confusion_matrix\n',confusion_matrix(y_test, y_pred))


confusion_matrix
 [[91 14]
 [17 57]]


In [233]:
print('classification_report\n',classification_report(y_test, y_pred))

classification_report
               precision    recall  f1-score   support

           0       0.84      0.87      0.85       105
           1       0.80      0.77      0.79        74

    accuracy                           0.83       179
   macro avg       0.82      0.82      0.82       179
weighted avg       0.83      0.83      0.83       179



# Feature importance

In [234]:
importance = pd.Series(model.feature_importances_, index=X.columns)

In [235]:
importance.sort_values(ascending=False)

Sex           0.269306
Fare          0.264815
Age           0.246012
Pclass        0.076197
FamilySize    0.051687
SibSp         0.029636
Embarked_S    0.023586
Parch         0.019965
Embarked_Q    0.009432
IsAlone       0.009364
dtype: float64

## Insights
- Gender is the strongest predictor of survival, aligning with the ‘women and children first’ evacuation policy.
- Ticket fare plays a major role, likely acting as a proxy for socioeconomic status
- Pclass is less important than expected, Fare already captures similar info
- 

# Comparing model with logistic regression

In [236]:
from sklearn.linear_model import LogisticRegression

In [237]:
lr=LogisticRegression(max_iter=1000)
lr.fit(X_train, y_train)
y_pred_lr=lr.predict(X_test)

In [238]:
from sklearn.metrics import accuracy_score
print("Logistic Regression Accuracy",accuracy_score(y_test,y_pred_lr))

Logistic Regression Accuracy 0.7988826815642458


In [239]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred_lr)

array([[89, 16],
       [20, 54]], dtype=int64)

In [240]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred_lr))

              precision    recall  f1-score   support

           0       0.82      0.85      0.83       105
           1       0.77      0.73      0.75        74

    accuracy                           0.80       179
   macro avg       0.79      0.79      0.79       179
weighted avg       0.80      0.80      0.80       179



## Insight  
- Random Forest performs better, as it captures captures non-linear relationships
- Titanic data has interactions like:
Sex + Pclass
Age + Fare
- logistic regression is simpler so it has slightly lower performance

# Improve model

## Adding new feature

In [241]:
df['AgeGroup'] = pd.cut(df['Age'], bins=[0,12,18,60,100], labels=[0,1,2,3])

In [242]:
from sklearn.ensemble import RandomForestClassifier

model_random = RandomForestClassifier(
    n_estimators=300,
    max_depth=6,
    min_samples_split=5,
    class_weight='balanced',
    random_state=42
)

In [243]:
model_random.fit(X_train, y_train)

RandomForestClassifier(class_weight='balanced', max_depth=6,
                       min_samples_split=5, n_estimators=300, random_state=42)

In [244]:
y_pred_random_improve=model_random.predict(X_test)

In [245]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

print('Accuracy',accuracy_score(y_test, y_pred_random_improve))

Accuracy 0.8156424581005587


In [246]:
print('confusion_matrix\n',confusion_matrix(y_test, y_pred_random_improve))


confusion_matrix
 [[91 14]
 [19 55]]


In [247]:
print('classification_report\n',classification_report(y_test, y_pred_random_improve))

classification_report
               precision    recall  f1-score   support

           0       0.83      0.87      0.85       105
           1       0.80      0.74      0.77        74

    accuracy                           0.82       179
   macro avg       0.81      0.80      0.81       179
weighted avg       0.81      0.82      0.81       179



## Insight
- previous random forset model performed better

# Saving mdoel

In [253]:
import pickle
import os
import pickle

os.makedirs('../model', exist_ok=True)

pickle.dump(model, open('../model/titanic_model.pkl', 'wb'))